extensions/cpsection/language/model.py

```diff
@@ -81,49 +81,54 @@ def _initialize():
 
 
 def _write_i18n(lang_env, language_env):
-    path = os.path.join(os.environ.get('HOME'), '.i18n')
-    if not os.access(path, os.W_OK):
-        print _standard_msg
-        fd = open(path, 'w')
-        fd.write('LANG="%s"\n' % _default_lang)
-        fd.write('LANGUAGE="%s"\n' % _default_lang)
-        fd.close()
-    else:
-        fd = open(path, 'w')
-        fd.write('LANG="%s"\n' % lang_env)
-        fd.write('LANGUAGE="%s"\n' % language_env)
-        fd.close()
+    path = os.path.join(os.environ.get('HOME', ''), '.i18n')
+    # try avoid writing the file if the values didn't changed
+    lang_line = None
+    language_line = None
+    try:
+        with open(path) as fd:
+            for line in fd:
+                if line.startswith('LANG='):
+                    lang_line = line
+                elif line.startswith('LANGUAGE='):
+                    language_line = line
+    except:
+        pass
+
+    new_lang_line = 'LANG="%s"\n' % lang_env
+    new_language_line = 'LANGUAGE="%s"\n' % language_env
+
+    if lang_line == new_lang_line and language_line == new_language_line:
+        return
 
+    try:
+        with open(path, 'w') as fd:
+            fd.write(new_lang_line)
+            fd.write(new_language_line)
+            fd.flush()
+            # be sure all is flushed
+            os.fsync(fd)
+    except:
+        logging.exception('Error writing .i18n file')
 
-def get_languages():
-    path = os.path.join(os.environ.get('HOME', ''), '.i18n')
-    if not os.access(path, os.R_OK):
-        print _standard_msg
-        fd = open(path, 'w')
-        fd.write('LANG="%s"\n' % _default_lang)
-        fd.write('LANGUAGE="%s"\n' % _default_lang)
-        fd.close()
-        return [_default_lang]
-
-    fd = open(path, 'r')
-    lines = fd.readlines()
-    fd.close()
 
+def get_languages():
+    # read the env variables set in bin/sugar
     langlist = None
     lang = _default_lang
 
-    for line in lines:
-        if line.startswith('LANGUAGE='):
-            lang = line[9:].replace('"', '')
-            lang = lang.strip()
-            if lang.endswith('UTF-8'):
-                lang = lang.replace('UTF-8', 'utf8')
-            langlist = lang.split(':')
-        elif line.startswith('LANG='):
-            lang = line[5:].replace('"', '')
-            lang = lang.strip()
-            if lang.endswith('UTF-8'):
-                lang = lang.replace('UTF-8', 'utf8')
+    if 'LANGUAGE' in os.environ:
+        lang = os.environ['LANGUAGE']
+        lang = lang.strip()
+        if lang.endswith('UTF-8'):
+            lang = lang.replace('UTF-8', 'utf8')
+        langlist = lang.split(':')
+
+    if 'LANG' in os.environ:
+        lang = os.environ['LANG']
+        lang = lang.strip()
```

Comment:
Lines 127 to 131 seem remarkably similar to lines 120 to 124 (not 125), so perhaps a local function inside get_languages would ease reading.